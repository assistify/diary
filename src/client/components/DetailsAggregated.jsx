import React from 'react';
import { PropTypes } from 'prop-types';

import Heading from 'react-bulma-components/lib/components/heading';
import Section from 'react-bulma-components/lib/components/section';
import Box from 'react-bulma-components/lib/components/box';
import Container from 'react-bulma-components/lib/components/container';
import ActivityItems from './Activity';
import { memberReportType } from '../../models/memberReportType';

export default function DetailsAggregated(props) {
  const { teamReport } = props;

  const allPlannedItems = teamReport.reduce(
    (all, memberReport) => {
      const plannedBy = memberReport.statusKnown && memberReport.future.plannedItems
        ? memberReport.future.plannedItems.map(plannedItem => ({
          title: plannedItem.title,
          details: plannedItem.details,
          owners: [memberReport.username]
        }))
        : [];

      return all.concat(plannedBy);
    },
    []
  );

  const aggregatedItems = allPlannedItems.reduce((aggregated, item) => {
    const sameTitleItemIndex = aggregated.findIndex(aggregatedItem => aggregatedItem.title === item.title); // eslint-disable-line max-len
    if (sameTitleItemIndex >= 0) {
      aggregated[sameTitleItemIndex].owners.push(...item.owners);
    } else {
      aggregated.push(item);
    }
    return aggregated;
  },
  []);

  return (
    <Section className="c-future">

      <Heading size={3}>Was ansteht</Heading>

      {
        aggregatedItems.length > 0
          ? (
            <ActivityItems
              title="Geplante TÃ¤tigkeiten"
              list={aggregatedItems}
              className="next"
            />
          )
          : (
            <Box>
              <Container className="no-planned-activities">Bisher nichts geplant</Container>
            </Box>
          )
      }
    </Section>

  );
}

DetailsAggregated.propTypes = {
  teamReport: PropTypes.arrayOf(PropTypes.shape(memberReportType)).isRequired
};
