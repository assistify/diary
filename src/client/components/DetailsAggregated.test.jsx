import React from 'react';
import renderer from 'react-test-renderer';
import DetailsAggregated from './DetailsAggregated';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <DetailsAggregated
        teamReport={[
          {
            username: 'user1',
            future: {
              availability: 'irrelevant',
              plannedItems: [
                { title: 'planned 1', details: 'some details' }
              ]
            },
            past: {
              blockingItems: []
            }
          },
          {
            username: 'user2',
            future: {
              availability: 'irrelevant',
              plannedItems: [
                { title: 'planned 2', details: 'some details' }
              ]
            },
            past: {
              blockingItems: [
                { title: 'blocking 2', details: 'some details' }
              ]
            }
          }
        ]}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});


it('aggregates correctly', () => {
  const tree = renderer
    .create(
      <DetailsAggregated
        teamReport={[
          {
            username: 'user1',
            statusKnown: true,
            future: {
              availability: 'irrelevant',
              plannedItems: [
                { title: 'planned 1', details: 'some details' }
              ]
            },
            past: {
              blockingItems: []
            }
          },
          {
            username: 'user2',
            statusKnown: true,
            future: {
              availability: 'irrelevant',
              plannedItems: [
                { title: 'planned 1', details: 'some details' }
              ]
            },
            past: {
              blockingItems: []
            }
          }
        ]}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders a placeholder if nothing\'s planned', () => {
  const tree = renderer
    .create(
      <DetailsAggregated
        teamReport={[
          {
            username: 'user1',
            future: {
              availability: 'irrelevant',
              plannedItems: []
            },
            past: {
              blockingItems: []
            }
          },
          {
            username: 'user2',
            future: {
              availability: 'irrelevant',
              plannedItems: []
            },
            past: {
              blockingItems: []
            }
          }
        ]}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
