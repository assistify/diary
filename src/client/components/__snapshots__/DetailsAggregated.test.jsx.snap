// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aggregates correctly 1`] = `
<section
  className="section c-future"
  style={Object {}}
>
  <h1
    className="title is-3"
  >
    Was ansteht
  </h1>
  <div
    className="box c-activities next"
    style={Object {}}
  >
    <h1
      className="title is-6"
    >
      Geplante Tätigkeiten
    </h1>
    <ul>
      <li>
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "planned 1",
            }
          }
        />
        <span
          className="container"
          style={Object {}}
        >
          <span
            className="content"
            style={Object {}}
          >
             (
          </span>
          <span
            className="container"
            style={Object {}}
          >
            <a
              className="c-user"
              href="/direct/user1"
              rel="noopener noreferrer"
              target="_blank"
            >
              @user1
            </a>
            <span
              className="content"
              style={Object {}}
            >
              , 
            </span>
          </span>
          <span
            className="container"
            style={Object {}}
          >
            <a
              className="c-user"
              href="/direct/user2"
              rel="noopener noreferrer"
              target="_blank"
            >
              @user2
            </a>
          </span>
          <span
            className="content"
            style={Object {}}
          >
            )
          </span>
        </span>
      </li>
    </ul>
  </div>
</section>
`;

exports[`renders a placeholder if nothing's planned 1`] = `
<section
  className="section c-future"
  style={Object {}}
>
  <h1
    className="title is-3"
  >
    Was ansteht
  </h1>
  <div
    className="box"
    style={Object {}}
  >
    <div
      className="container no-planned-activities"
      style={Object {}}
    >
      Bisher nichts geplant
    </div>
  </div>
</section>
`;

exports[`renders correctly 1`] = `
<section
  className="section c-future"
  style={Object {}}
>
  <h1
    className="title is-3"
  >
    Was ansteht
  </h1>
  <div
    className="box"
    style={Object {}}
  >
    <div
      className="container no-planned-activities"
      style={Object {}}
    >
      Bisher nichts geplant
    </div>
  </div>
</section>
`;
